src/login/logind-session.c

```diff
@@ -1242,6 +1252,14 @@ bool session_may_gc(Session *s, bool drop_not_started) {
                         return false;
         }
 
+        if (s->leader.fd >= 0) {
+                r = pidfd_is_alive(s->leader.fd);
```

Comment:
so I'd like you to add a pidref_is_alive() wrapper on top of your new pidfd_is_alive() that uses pidfd_is_alive() if we have a pidfd, and otherwise uses pid_is_alive().

Then just call this here, generically.